sunscreen/src/lib.rs

```diff
@@ -105,39 +110,104 @@ impl Application {
-        if programs.is_empty() {
+    pub(crate) fn new(
+        fhe_programs: HashMap<String, CompiledFheProgram>,
+        zkp_programs: HashMap<String, ZkpFrontendCompilation>,
+    ) -> Result<Self> {
+        if fhe_programs.is_empty() && zkp_programs.is_empty() {
             return Err(Error::NoPrograms);
         }
 
-        Ok(Self { programs })
+        Ok(Self {
+            fhe_programs,
+            zkp_programs,
+            _phantom: PhantomData,
+        })
     }
+}
 
+impl<T> Application<T>
+where
+    T: marker::Fhe,
+{
     /**
      * Returns the [`Params`] suitable for running each contained [`CompiledFheProgram`].
      * These parameters were chosen during compilation.
+     *
+     * # Remarks
+     * If no [`fhe_program`] was specified, this function returns [`None`].
      */
     pub fn params(&self) -> &Params {
-        // We can safely unwrap the iterator because we ensured we have at
-        // least 1 program during construction.
-        &self.programs.values().next().unwrap().metadata.params
+        &self.fhe_programs.values().next().unwrap().metadata.params
     }
 
+    #[deprecated]
     /**
      * Gets the [`CompiledFheProgram`] with the given name or [`None`] if not present.
      */
     pub fn get_program<N>(&self, name: N) -> Option<&CompiledFheProgram>
     where
         N: AsRef<str>,
     {
-        self.programs.get(name.as_ref())
+        self.get_fhe_program(name)
     }
 
+    /**
+     * Gets the [`CompiledFheProgram`] with the given name or [`None`] if not present.
+     */
+    pub fn get_fhe_program<N>(&self, name: N) -> Option<&CompiledFheProgram>
+    where
+        N: AsRef<str>,
+    {
+        self.fhe_programs.get(name.as_ref())
+    }
+
+    #[deprecated]
     /**
      * Returns an iterator over all the compiled programs.
+     *
+     * # Deprecated
+     * Please use [`get_fhe_programs`](Self::get_fhe_programs) instead.
      */
     pub fn get_programs(&self) -> impl Iterator<Item = (&String, &CompiledFheProgram)> {
-        self.programs.iter()
+        self.get_fhe_programs()
+    }
+
+    /**
+     * Returns an iterator over all the compiled programs.
+     */
+    pub fn get_fhe_programs(&self) -> impl Iterator<Item = (&String, &CompiledFheProgram)> {
+        self.fhe_programs.iter()
+    }
+}
+
+impl Application<Zkp> {
+    pub(crate) fn to_application(other: Application<()>) -> Application<Zkp> {
+        Self {
+            fhe_programs: other.fhe_programs,
+            zkp_programs: other.zkp_programs,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl Application<Fhe> {
+    pub(crate) fn to_application(other: Application<()>) -> Application<Fhe> {
+        Self {
+            fhe_programs: other.fhe_programs,
+            zkp_programs: other.zkp_programs,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl Application<FheZkp> {
+    pub(crate) fn to_application(other: Application<()>) -> Application<FheZkp> {
```

Comment:
These impls seem odd. It seems like anytime someone uses `Application::new()` they get back an `app: Application<()>` and then need to use `Application<Something>::to_application(app)`? I imagine we can simplify this. 

In general I wonder if these `Struct<()>` impls with `new` and whatnot could instead just be generic. Then someone can specify what they want by just specifying the generic, e.g. `Application<Fhe>::new().get_fhe_program()`.

Reply:
This should be an `into()` function as we did with Compiler. But the idea is the same: our code internally needs to convert between types when appropriate, but users never do.

Users never construct Applications themselves (the `new` method is marked pub(crate)). Rather, the compiler emits the appropriately typed one.

Reply:
I've migrated to the same pattern as into with a method on

```
impl<T> Application<T> {
  fn convert<U>(self) -> Application<U> { ... }
}
```